ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestinationSync.java

```diff
@@ -598,13 +624,19 @@ public Map<String, CloudBlobMetadata> getBlobMetadata(List<BlobId> blobIds) thro
       try {
         BlobProperties blobProperties = getBlobProperties(blobLayout);
         cloudBlobMetadataMap.put(blobId.getID(), CloudBlobMetadata.fromMap(blobProperties.getMetadata()));
-      } catch (Throwable t) {
-        if (t instanceof CloudStorageException
-            && ((CloudStorageException) t).getStatusCode() == CloudBlobStore.STATUS_NOT_FOUND) {
-          // We should never be here because replication logic checks if a blob exists or not before undeleting it.
-          String dbg = String.format("Blob %s absent in Azure blob storage", blobLayout);
-          logger.error(dbg);
+      } catch (CloudStorageException cse) {
+        if (cse.getCause() instanceof BlobStorageException &&
+            ((BlobStorageException) cse.getCause()).getErrorCode() == BlobErrorCode.BLOB_NOT_FOUND) {
+          /*
+            We mostly get here from findMissingKeys, and we will encounter many blobs missing from cloud before we upload them.
```

Comment:
I agree with you that this is happening because findMissingKeys() is using `getBlobMetadata` instead of `exists()` to check if blob is present or that. But, I just feel that ignoring `BLOB_NOT_FOUND` exception here assuming that it is always called from `findMissingKeys()` may be not correct. If we look at this method independently, `BLOB_NOT_FOUND` seems to be a valid exception (For example, if this method is used to get blob metadata in recovery path, this exception could be valid).

If it is flooding the logging, would it be possible (in future, if not now) to change the log type to `trace` in this method and log the `error` in caller methods?

Reply:
`getBlobProperties` logs the trace message for blob-not-found. I can log it here in `getBlobMetadata` as well but it'd just be a duplicate.  However, I cannot throw the blob-not-found err as this fn accepts a list of blobs and must loop through the entire list. 

Overall, the design and structure of current code is restricted by the choices made by previous authors. There is quite a bit inconsistency. Some fns accept a list while others accept a single item. Some return boolean, others int and others void.